core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -180,33 +180,15 @@ private[zio] final class FiberContext[E, A](
                 case IO.Tags.AsyncEffect =>
                   val io = curIo.asInstanceOf[IO.AsyncEffect[E, Any]]
 
-                  // By default, we'll resume asynchronously, so this
-                  // evaluation loop should halt:
-                  curIo = null
-
                   // Enter suspended state:
-                  val cancel = enterAsync()
-
-                  if (cancel ne null) {
-                    // It's possible we were interrupted prior to entering
-                    // suspended state. So we have to check that condition,
-                    // and if so, do not initiate the async effect (because
-                    // otherwise, it would not be interrupted).
-                    try {
-                      if (shouldDie) curIo = IO.interrupt
-                      else
-                        io.register(resumeAsync) match {
-                          case Async.Now(value) =>
-                            if (shouldResumeAsync()) curIo = IO.done(value)
-
-                          case Async.MaybeLater(cancel0) => cancel.set(cancel0)
-                        }
-                    } finally {
-                      // May not allow exit of the code block without the
-                      // cancel action being set (could hang interruptor!):
-                      if (!cancel.isSet) cancel.set(IO.unit)
+                  if (enterAsync()) {
+                    io.register(resumeAsync) match {
+                      case Async.Now(io) =>
+                        if (exitAsync()) curIo = io
```

Comment:
I'm always wondering about the absence of an `else` clause here: shouldn't it be
```scala
curIo = if (exitAsync()) io else null
```

Reply:
❤️ Beautiful! Wow, that cleaned up this code a lot and provides stronger guarantees on handling all cases...